server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProviderFactory.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.userprofile;
+
+import org.keycloak.provider.ProviderFactory;
+
+/**
+ * @author <a href="mailto:markus.till@bosch.io">Markus Till</a>
+ */
+public interface UserProfileProviderFactory extends  ProviderFactory<UserProfileProvider> {
```

Comment:
I think we could have some methods here to create the `UserProfileProvider` instance without pushing too much to clients of the API. For instance:

```
Map<String, List<String>> newAttributes = // obtain attributes to update
UserProfileProviderFactory factory = // obtain factory from session
UserProfileProvider provider = factory.create(UserUpdateEvent.IdpReview, user.getAttributes(), newAttributes);
UserProfileValidationResult result = profileProvider.validate();
```

This can hide some implementation details such as the `UserProfileContext` implementation that should be used.

In addition to that, the user ID also works like any other attribute into the attribute map that is the real source of data and data structure we care about to perform the validations. 

The `Attributes` API I gave as an example could also help to make this SPI more attribute oriented and simplify even more how clients pass over the attributes from the different places in Keycloak as well as using different sources.

Reply:
@pedroigor  I tried to go into this direction. It's not possible to do this on the SPI level as the interfaces would have to know the concrete types in the service impl. So I moved the factory like methods into the DefaultUserProfileContext